model/host/db.go

```diff
@@ -1474,3 +1480,60 @@ func FindUnexpirableRunningWithoutPersistentDNSName(ctx context.Context, limit i
 		PersistentDNSNameKey: nil,
 	}, options.Find().SetLimit(int64(limit)))
 }
+
+// FindHostsScheduledToStop finds all unepxirable hosts that are due to stop due to their sleep
+// schedule settings.
+func FindHostsScheduledToStop(ctx context.Context) ([]Host, error) {
+	now := time.Now()
+	sleepScheduleNextStopKey := bsonutil.GetDottedKeyName(SleepScheduleKey, SleepScheduleNextStopTimeKey)
+	sleepSchedulePermanentlyExemptKey := bsonutil.GetDottedKeyName(SleepScheduleKey, SleepSchedulePermanentlyExemptKey)
+	sleepScheduleTemporarilyExemptUntil := bsonutil.GetDottedKeyName(SleepScheduleKey, SleepScheduleTemporarilyExemptUntilKey)
+	sleepScheduleShouldKeepOff := bsonutil.GetDottedKeyName(SleepScheduleKey, SleepScheduleShouldKeepOffKey)
+
+	return Find(ctx, bson.M{
+		StatusKey:                         bson.M{"$in": []string{evergreen.HostRunning, evergreen.HostStopping}},
+		StartedByKey:                      bson.M{"$ne": evergreen.User},
+		NoExpirationKey:                   true,
+		sleepScheduleNextStopKey:          bson.M{"$lte": now},
+		sleepSchedulePermanentlyExemptKey: bson.M{"$ne": true},
+		"$or": []bson.M{
+			{
+				sleepScheduleTemporarilyExemptUntil: nil,
+			},
+			{
+				sleepScheduleTemporarilyExemptUntil: bson.M{"$lte": now},
+			},
+		},
+		sleepScheduleShouldKeepOff: bson.M{"$ne": true},
```

Comment:
Wouldn't we want to stop a host that's marked to be kept off?

Reply:
The keep off flag isn't implemented yet, but I was envisioning that the sleep schedule itself wouldn't have to do anything extra, because the host should already be off when it's set to keep off. The flag is supposed to be set when the user stops the host (as an optional toggle), so they would press pause in the UI, then the host would stop and if they selected it, it would stay off until they turn it back on. Once they turn their host back on, Evergreen would clear the flag so that it resumes the usual sleep schedule. So it seems like the background sleep schedule shouldn't need to do anything extra to handle it.

Reply:
But this query also only matches on things that are in running or stopping. It seems like not including keep off here could help us fix a state, if it gets wonky (and/or just doesn't really need to be in the query at all, if we're certain the status won't be running/stopping in this case).

Notably I don't feel strongly about this, it just seems like we don't need it.

Reply:
It makes the calculations kind of weird because if it's not stopped, then the next stop/start time are probably wrong, so I don't really want to add extra logic to prevent it from starting back up. I think it's okay, we can revisit if we find issues hosts where aren't off when should keep off is enabled.